cdisutils/storage.py

```diff
@@ -8,25 +8,118 @@
 from future.standard_library import install_aliases
 install_aliases()
 
-from urllib.parse import urlparse
-
-import hashlib
-import json
-import re
+from cdisutils.log import get_logger
+from cdisutils.parsers import S3URLParser
 
 # TODO: add tests
 
 
+class S3ConnectionProxyFix(S3Connection):
+    """Custom S3Connection class to workaround boto proxy issue.
+
+    This fix will apply to any Python 3 boto code that attempts to interact with
+    Amazon S3 because it needs a proxy to access it.
+
+    With the Python 3 upgrade came breaking changes specifically pertaining to
+    bytes and strings. The sendall() socket command no longer accepts strings.
+    We have to convert them all to bytes before sending.
+
+    There have been attempts to merge a fix for this into the main boto branch
+    but they have not been successful.
+    - https://github.com/boto/boto/pull/2718
+    - https://github.com/boto/boto/pull/3699
+    """
+    def __init__(self, *args, **kwargs):
+        super(S3ConnectionProxyFix, self).__init__(*args, **kwargs)
+        self.logger = get_logger("S3ConnectionProxyFix")
+
+    def proxy_ssl(self, host=None, port=None):
+        if host and port:
+            host = '%s:%d' % (host, port)
+        else:
+            host = '%s:%d' % (self.host, self.port)
+        # Seems properly to use timeout for connect too
+        timeout = self.http_connection_kwargs.get("timeout")
+        if timeout is not None:
+            sock = socket.create_connection((self.proxy,
+                                             int(self.proxy_port)), timeout)
+        else:
+            sock = socket.create_connection((self.proxy, int(self.proxy_port)))
+        self.logger.debug("Proxy connection: CONNECT %s HTTP/1.0\r\n", host)
+        sock.sendall(("CONNECT %s HTTP/1.0\r\n" % host).encode())
+        sock.sendall(("User-Agent: %s\r\n" % UserAgent).encode())
+        if self.proxy_user and self.proxy_pass:
+            for k, v in iteritems(self.get_proxy_auth_header()):
+                sock.sendall(("%s: %s\r\n" % (k, v)).encode())
+            # See discussion about this config option at
+            # https://groups.google.com/forum/?fromgroups#!topic/boto-dev/teenFvOq2Cc
+            if config.getbool('Boto', 'send_crlf_after_proxy_auth_headers', False):
+                sock.sendall(b"\r\n")
+        else:
+            sock.sendall(b"\r\n")
+        resp = http_client.HTTPResponse(sock, debuglevel=self.debug)
+        resp.begin()
+
+        if resp.status != 200:
+            # Fake a socket error, use a code that make it obvious it hasn't
+            # been generated by the socket library
+            raise socket.error(-71,
+                               "Error talking to HTTP proxy %s:%s: %s (%s)" %
+                               (self.proxy, self.proxy_port,
+                                resp.status, resp.reason))
+
+        # We can safely close the response, it duped the original socket
+        resp.close()
```

Comment:
I believe we are closing a file pointer with this. We should make sure to put the above logic into a try block and this close into a finally block.